feat: configurable html elements and attributes

david-abell opened the pull request:
I was very happy to find your package a couple of days ago. 

This pull request adds parcel's config load method to point to an optional JSON configuration file in a user's project root directory. The json configuration file may contain key: value pairs of HTML elements and custom data attributes `either as a string or an array of strings` for the basic `addURLDependency()` method.

If an array is specified in the configuration json, it will only change the first found custom data-attribute per HTML element 

See the added `mock/html-datasrc.config.json` for configuration

I unfortunately have not learned how to write tests yet to add those as well.

david-abell committed:
add json configuration for custom url properties

david-abell committed:
remove loop

david-abell committed:
fix config load and prop comparison logic

david-abell committed:
add optional arrays to custom HTML element values

david-abell committed:
remove test package name

moritzlaube reviewed (changes requested):
Thanks so much for your input on this, @david-abell! I think, this is a marvelous idea. Especially, since you have a working default if you're not providing any config file. The only thing that I maybe would try to implement is the ability to also parse attributes with multiple width or pixel density descriptors like in the srcset-attribute.  If you'd have an idea how to do it I'm happy to write a test for it. Otherwise I would give it a try myself on monday. Also, just a few moments before you're pull request I updated the package to also parse picture and source tags. Thanks again for your contribution!

moritzlaube commented:
On the other hand. If you need the srcset-attribute, you'll use the `data-srcset` attribute for lazy-loading. Any other custom data-attribute will most probably only have one value like `data-bg-hidpi="lazy@2x.jpg"` in the `vanilla-lazyload` package...

But this very package also has this syntax: 

```
data-bg-multi="url(lazy-head.jpg), 
    url(lazy-body.jpg), 
    linear-gradient(#fff, #ccc)"
 ```
 This wouldn't work currently, since the image-paths are wrapped in `url()`.

david-abell committed:
outlined if statements

david-abell committed:
add function for combined url and background attrs

david-abell commented:
Good suggestions. I had not noticed that combined syntax which obviously breaks things. 

I have a working function to handle attributes like the `data-bg-multi`, although not yet for the lozad version as the format there wraps the link inside `url(/file)` are also wrapped in `''` like below:

`data-background-image-set="url('photo.jpg') 1x, url('photo@2x.jpg') 2x"`

 I believe the updated version of your map function below may actually handle everything else, at least with some preliminary tests. All it needed was the null check on width!

      if ((tag === 'img' || tag === 'source') && attrs['data-srcset'] != null) {
        const srcsets = attrs['data-srcset'].split(',').map(item => {
          const [url, width] = item.trim().split(' ')
          if (!width) return `${asset.addURLDependency(url)}`
          return `${asset.addURLDependency(url)} ${width}`
        })
        attrs['data-srcset'] = srcsets.join(', ')
        isDirty = true
      }

Since the complexity goes way up because of the combined attributes like It `data-bg-multi` it makes the most sense to me just to define the supported tag and attributes in one spot and then pass them to the right function. Also super easy to add basic `tag: attr` pairs later.

I'm working to combines that initial config along with the user config import that I originally submitted this for. I like where it is going so far though.

david-abell committed:
add ramda, merge configs and depency functions

david-abell committed:
remove unused code

david-abell committed:
remove log statements

david-abell committed:
Merge branch 'main' into customurlprop

david-abell committed:
Merge branch 'customurlpropfordescriptors' into customurlprop
